File: apps/web/src/lib/hooks/voice-session/transcript-safety.ts
/**
 * Voice Transcript Safety Checker
 * Validates user transcript text from voice sessions for content safety
 *
 * Related: Plan 148 W2 T2-04 - User transcript safety check
 * Compliance: VCE-002 logging checkpoint (EU AI Act Art. 13, 29)
 */

'use client';

import { filterInput, detectJailbreak } from '@/lib/safety';
import type { FilterSeverity, FilterAction } from '@/lib/safety';
import { isFeatureEnabled } from '@/lib/feature-flags/client';
import { clientLogger as logger } from '@/lib/logger/client';

/**
 * VCE-002 compliant safety check result
 * Maps to compliance logging checkpoint from plan-147-notes.md
 */
export interface TranscriptSafetyResult {
  /** Detected severity level */
  severity: FilterSeverity;
  /** Flagged pattern categories */
  flaggedPatterns: string[];
  /** Action to take based on safety assessment */
  actionTaken: 'allow' | 'warn' | 'block' | 'escalate';
  /** Duration of safety check in milliseconds */
  checkDurationMs: number;
}

/**
 * Check user transcript for safety violations
 * Guards behind voice_transcript_safety feature flag
 *
 * @param sessionId - Voice session ID for logging correlation
 * @param transcriptText - User's transcribed speech
 * @returns VCE-002 compliant safety result
 *
 * @example
 * const result = checkUserTranscript(sessionId, 'Can you help with math?');
 * if (result.actionTaken === 'block') {
 *   // Trigger safety intervention (T2-06)
 * }
 */
export function checkUserTranscript(
  sessionId: string,
  transcriptText: string,
): TranscriptSafetyResult {
  const startTime = performance.now();

  // Check feature flag
  const flagResult = isFeatureEnabled('voice_transcript_safety');
  if (!flagResult.enabled) {
    logger.debug('[TranscriptSafety] Feature disabled, skipping check', { sessionId });
    return {
      severity: 'none',
      flaggedPatterns: [],
      actionTaken: 'allow',
      checkDurationMs: performance.now() - startTime,
    };
  }

  // Early return for empty input
  const normalized = transcriptText.trim();
  if (!normalized) {
    return {
      severity: 'none',
      flaggedPatterns: [],
      actionTaken: 'allow',
      checkDurationMs: performance.now() - startTime,
    };
  }

  // Run content filter
  const filterResult = filterInput(transcriptText);

  // Map filter action to VCE-002 action
  let actionTaken = mapFilterAction(filterResult.action);
  let severity = filterResult.severity;

  // Extract flagged patterns
  const flaggedPatterns: string[] = [];
  if (filterResult.category) {
    flaggedPatterns.push(filterResult.category);
  }

  // T1.5: Advanced jailbreak / prompt-injection detection on the transcript.
  // filterInput above already flags obvious JAILBREAK_PATTERNS as
  // category 'jailbreak'; the dedicated detector catches sophisticated
  // attempts the regex misses (encoding, code injection, crescendo). Gate on
  // the detector's own action (block/terminate_session, i.e. threat >= high),
  // matching the chat/stream routes. When it fires we MUST elevate actionTaken
  // and severity — not just append the pattern — because event-handlers only
  // triggers the intervention (and triggerSafetyIntervention only acts) when
  // actionTaken !== 'allow'. getRedirectMessage already has a 'jailbreak'
  // branch, so the spoken redirect is correct once 'jailbreak' is flagged.
  // NOTE: single-transcript detection only (no conversation history is passed
  // here), so multi_turn/crescendo buildup across turns is not detected on the
  // voice path — a known limitation vs the chat routes.
  const jailbreakResult = detectJailbreak(transcriptText);
  if (jailbreakResult.action === 'block' || jailbreakResult.action === 'terminate_session') {
    if (!flaggedPatterns.includes('jailbreak')) {
      flaggedPatterns.push('jailbreak');
    }
    // Elevate action: block -> 'block', terminate -> 'escalate'. Never
    // downgrade an already-stronger action from filterInput (e.g. crisis
    // 'escalate').
    const elevated = jailbreakResult.action === 'terminate_session' ? 'escalate' : 'block';
    if (actionTaken !== 'escalate') {
      actionTaken = elevated;
    }
    // Elevate severity to the detector's threat level (high/critical) when it
    // is stronger than the content-filter severity.
    const rank: Record<FilterSeverity, number> = {
      none: 0,
      low: 1,
      medium: 2,
      high: 3,
      critical: 4,
    };
    if (rank[jailbreakResult.threatLevel] > rank[severity]) {
      severity = jailbreakResult.threatLevel;
    }
  }

  const checkDurationMs = performance.now() - startTime;

  const result: TranscriptSafetyResult = {
    severity,
    flaggedPatterns,
    actionTaken,
    checkDurationMs,
  };

  // Log safety check (VCE-002 checkpoint)
  if (actionTaken !== 'allow') {
    logger.info('[TranscriptSafety] Safety check flagged user transcript', {
      component: 'voice-transcript-safety',
      eventId: 'VCE-002',
      eventName: 'Transcript Safety Check (Input)',
      sessionId,
      detectedSeverity: result.severity,
      flaggedPatterns: result.flaggedPatterns,
      actionTaken: result.actionTaken,
      checkDurationMs: result.checkDurationMs,
      // Do NOT log transcriptText in production (GDPR Art. 25)
    });
  } else {
    logger.debug('[TranscriptSafety] User transcript passed safety check', {
      sessionId,
      checkDurationMs: result.checkDurationMs,
    });
  }

  return result;
}

/**
 * VCE-003 compliant assistant transcript safety check result
 * Maps to compliance logging checkpoint from plan-147-notes.md
 */
export interface AssistantTranscriptSafetyResult {
  /** Detected severity level */
  severity: FilterSeverity;
  /** Flagged pattern categories */
  flaggedPatterns: string[];
  /** Action to take based on safety assessment */
  actionTaken: 'allow' | 'sanitize' | 'reject';
  /** Duration of safety check in milliseconds */
  checkDurationMs: number;
}

/**
 * Check assistant transcript for safety violations
 * Guards behind voice_transcript_safety feature flag
 *
 * @param sessionId - Voice session ID for logging correlation
 * @param transcriptText - Assistant's generated speech
 * @returns VCE-003 compliant safety result
 *
 * @example
 * const result = checkAssistantTranscript(sessionId, 'Let me help you with that problem.');
 * if (result.actionTaken === 'reject') {
 *   // Log to audit, escalate for review
 * }
 */
export function checkAssistantTranscript(
  sessionId: string,
  transcriptText: string,
): AssistantTranscriptSafetyResult {
  const startTime = performance.now();

  // Check feature flag
  const flagResult = isFeatureEnabled('voice_transcript_safety');
  if (!flagResult.enabled) {
    logger.debug('[TranscriptSafety] Feature disabled, skipping assistant check', { sessionId });
    return {
      severity: 'none',
      flaggedPatterns: [],
      actionTaken: 'allow',
      checkDurationMs: performance.now() - startTime,
    };
  }

  // Early return for empty input
  const normalized = transcriptText.trim();
  if (!normalized) {
    return {
      severity: 'none',
      flaggedPatterns: [],
      actionTaken: 'allow',
      checkDurationMs: performance.now() - startTime,
    };
  }

  // Run content filter on assistant output
  const filterResult = filterInput(transcriptText);
  const checkDurationMs = performance.now() - startTime;

  // Map filter action to VCE-003 action taxonomy
  const actionTaken = mapAssistantFilterAction(filterResult.action);

  // Extract flagged patterns
  const flaggedPatterns: string[] = [];
  if (filterResult.category) {
    flaggedPatterns.push(filterResult.category);
  }

  const result: AssistantTranscriptSafetyResult = {
    severity: filterResult.severity,
    flaggedPatterns,
    actionTaken,
    checkDurationMs,
  };

  // Log safety check (VCE-003 checkpoint). Filter firing means the safety
  // system is working as designed — log at info to keep Sentry noise low
  // while preserving structured metadata for audit trails (MIRRORBUDDY-1Q).
  if (actionTaken !== 'allow') {
    logger.info('[TranscriptSafety] Safety filter applied to assistant transcript', {
      component: 'voice-transcript-safety',
      eventId: 'VCE-003',
      eventName: 'Transcript Safety Check (Output)',
      sessionId,
      detectedSeverity: result.severity,
      flaggedPatterns: result.flaggedPatterns,
      actionTaken: result.actionTaken,
      checkDurationMs: result.checkDurationMs,
      // Do NOT log transcriptText in production (GDPR Art. 25)
    });
  } else {
    logger.debug('[TranscriptSafety] Assistant transcript passed safety check', {
      sessionId,
      checkDurationMs: result.checkDurationMs,
    });
  }

  return result;
}

/**
 * Map content filter action to VCE-002 action taxonomy
 * Aligns with compliance logging checkpoint from plan-147-notes.md
 */
function mapFilterAction(filterAction: FilterAction): 'allow' | 'warn' | 'block' | 'escalate' {
  switch (filterAction) {
    case 'allow':
      return 'allow';
    case 'warn':
      return 'warn';
    case 'block':
      return 'block';
    case 'redirect':
      // redirect (crisis) escalates to human oversight
      return 'escalate';
    default:
      return 'block';
  }
}

/**
 * Map content filter action to VCE-003 action taxonomy for assistant output
 * Assistant violations are treated more strictly: reject vs sanitize
 */
function mapAssistantFilterAction(filterAction: FilterAction): 'allow' | 'sanitize' | 'reject' {
  switch (filterAction) {
    case 'allow':
      return 'allow';
    case 'warn':
      // Warnings on assistant output -> sanitize (log but allow)
      return 'sanitize';
    case 'block':
    case 'redirect':
      // Assistant should never produce content requiring block/redirect
      return 'reject';
    default:
      return 'reject';
  }
}

File: robot/reachy_mini_mirrorbuddy/transcript_safety.py
"""Pure-Python port of the web voice transcript safety contract.

Always enabled, matching the web default; no feature-flag disable path. User
checks include single-turn jailbreak detection, never multi-turn history.
Assistant ``sanitize`` means log but allow, not text rewriting. Preserve web
ordering and known substring false positives rather than inventing new policy.
Runtime depends only on the standard library; generated constants ship in-package.
"""

from collections.abc import Iterable
from dataclasses import dataclass
import logging
from time import perf_counter

from . import transcript_safety_patterns as patterns
from .transcript_safety_jailbreak import jailbreak_score
from .transcript_safety_redirects import REDIRECTS
from .transcript_safety_regex import JS_SPACE, compile_js, js_string

logger = logging.getLogger(__name__)


@dataclass(frozen=True, slots=True)
class TranscriptSafetyResult:
    severity: str
    flagged_patterns: tuple[str, ...]
    action_taken: str
    check_duration_ms: float


_CRISIS = tuple(compile_js(*spec) for spec in patterns.CRISIS_PATTERNS)
_PII = tuple(compile_js(*spec) for spec in patterns.PII_PATTERNS)
_FILTERS = tuple(
    (tuple(compile_js(*spec) for spec in specs), severity, action, category)
    for specs, severity, action, category in (
        (patterns.VIOLENCE_PATTERNS, "high", "block", "violence"),
        (patterns.HACKING_PATTERNS, "high", "block", "violence"),
        (patterns.JAILBREAK_PATTERNS, "high", "redirect", "jailbreak"),
        (patterns.EXPLICIT_PATTERNS, "high", "block", "explicit"),
        (patterns.PROFANITY_IT, "medium", "warn", "profanity"),
        (patterns.PROFANITY_EN, "medium", "warn", "profanity"),
    )
)
_RANK = {"none": 0, "low": 1, "medium": 2, "high": 3, "critical": 4}


def _filter_input(text: str) -> tuple[str, str, tuple[str, ...]]:
    lower = js_string(text.lower())
    normalized = lower.strip(JS_SPACE)
    if not normalized:
        return "none", "allow", ()
    if next((True for pattern in _CRISIS if pattern.search(lower)), False):
        return "critical", "redirect", ("crisis",)
    for expressions, severity, action, category in _FILTERS:
        if next((True for pattern in expressions if pattern.search(normalized)), False):
            return severity, action, (category,)
    if next((True for word in patterns.SEVERE if word in normalized), False):
        return "high", "block", ("violence",)
    original = js_string(text)
    if next((True for pattern in _PII if pattern.search(original)), False):
        return "medium", "block", ("pii",)
    return "none", "allow", ()


def _result(start: float, severity: str, flagged: tuple[str, ...], action: str,
            assistant: bool) -> TranscriptSafetyResult:
    result = TranscriptSafetyResult(severity, flagged, action, (perf_counter() - start) * 1000)
    if action != "allow":
        logger.info(
            "Safety filter applied to %s transcript", "assistant" if assistant else "user",
            extra={
                "component": "voice-transcript-safety",
                "eventId": "VCE-003" if assistant else "VCE-002",
                "eventName": "Transcript Safety Check (Output)" if assistant
                else "Transcript Safety Check (Input)",
                "detectedSeverity": result.severity,
                "flaggedPatterns": result.flagged_patterns,
                "actionTaken": result.action_taken,
                "checkDurationMs": result.check_duration_ms,
            },
        )
    return result


def check_user_transcript(text: str | None) -> TranscriptSafetyResult:
    """Check user input, elevating only high/critical advanced jailbreak scores."""
    start = perf_counter()
    text = text or ""
    severity, action, flagged = _filter_input(text)
    action = "escalate" if action == "redirect" else action
    if text.strip(JS_SPACE):
        score = jailbreak_score(text)
        if score >= 0.7:
            if "jailbreak" not in flagged:
                flagged += ("jailbreak",)
            elevated = "escalate" if score >= 0.9 else "block"
            if action != "escalate":
                action = elevated
            threat = "critical" if score >= 0.9 else "high"
            if _RANK[threat] > _RANK[severity]:
                severity = threat
    return _result(start, severity, flagged, action, False)


def check_assistant_transcript(text: str | None) -> TranscriptSafetyResult:
    """Check assistant output with web allow/sanitize/reject action taxonomy."""
    start = perf_counter()
    severity, action, flagged = _filter_input(text or "")
    mapped = {"allow": "allow", "warn": "sanitize", "block": "reject", "redirect": "reject"}
    return _result(start, severity, flagged, mapped[action], True)


def get_redirect_message(flagged_patterns: Iterable[str] | None) -> str:
    """Return the exact Italian web redirect, including its priority order."""
    flagged = tuple(flagged_patterns or ())
    for category in ("violence", "crisis", "explicit", "bias", "jailbreak", "profanity"):
        if category in flagged or (category == "explicit" and "age_inappropriate" in flagged):
            return REDIRECTS[category]
    return REDIRECTS["default"]
